lms/views/basic_lti_launch.py

```diff
@@ -0,0 +1,203 @@
+    add_user_to_group,
+    report_lti_launch,
+)
+
+
+@view_defaults(
+    decorator=[
+        # Before any LTI assignment launch create or update the Hypothesis user
+        # and group corresponding to the LTI user and course.
+        upsert_h_user,
+        upsert_course_group,
+        add_user_to_group,
+        # Report all LTI assignment launches to the /reports page.
+        report_lti_launch,
+    ],
+    feature_flag="new_oauth",
+    permission="launch_lti_assignment",
+    renderer="lms:templates/basic_lti_launch/basic_lti_launch.html.jinja2",
+    request_method="POST",
+    route_name="lti_launches",
+)
+class BasicLTILaunchViews:
+    def __init__(self, request):
+        self.request = request
+
+    @view_config(
+        canvas_file=True,
+        renderer="lms:templates/basic_lti_launch/canvas_file_basic_lti_launch.html.jinja2",
+    )
+    def canvas_file_basic_lti_launch(self):
+        """
+        Respond to a Canvas file assignment launch.
+
+        Canvas file assignment launch requests have a ``file_id`` request
+        parameter, which is the Canvas instance's ID for the file. To display
+        the assignment we have to use this ``file_id`` to get a download URL
+        for the file from the Canvas API. We then pass that download URL to
+        Via. We have to re-do this file-ID-for-download-URL exchange on every
+        single launch because Canvas's download URLs are temporary.
+        """
+        canvas_api_client = self.request.find_service(name="canvas_api_client")
+        file_id = self.request.params["file_id"]
+
+        try:
+            document_url = canvas_api_client.public_url(file_id)
+        except CanvasAPIError:
+            return {}
+
+        return {"via_url": via_url(self.request, document_url)}
+
+    @view_config(db_configured=True)
+    def db_configured_basic_lti_launch(self):
+        """
+        Respond to a DB-configured assignment launch.
+
+        DB-configured assignment launch requests don't have any kind of file ID
+        or document URL in the request. Instead the document URL is stored in
+        our own DB. This happens with LMS's that don't support LTI content item
+        selection/deep linking, so they don't support storing the document URL
+        in the LMS and passing it back to us in each launch request. Instead we
+        retrieve the document URL from the DB and pass it to Via.
+        """
+        resource_link_id = self.request.params["resource_link_id"]
+        tool_consumer_instance_guid = self.request.params["tool_consumer_instance_guid"]
+
+        # The ``db_configured=True`` view predicate ensures that this view
+        # won't be called if there isn't a matching ModuleItemConfiguration in
+        # the DB. So here we can safely assume that the ModuleItemConfiguration
+        # exists.
+        document_url = ModuleItemConfiguration.get_document_url(
+            self.request.db, tool_consumer_instance_guid, resource_link_id
+        )
```

Comment:
This pattern has been serving me well in the LMS app: service and model methods that just use strings (and other basic types or, if necessary, values like `LTIUser`) as arguments and return values. Rather than the calling code ever having to handle model objects themselves. It's usually the case that a model object isn't actually needed, just one or two fields.